### DISCUSSION TITLE
Alist V3版本API文档- 个人整理

### DISCUSSION THREAD

**Discussion by Kuingsmile** [AlistGo/alist#2501] [Show and tell]
个人整理了AList V3版本的一些常用API和使用方法，可见下方链接和文件下载。

## 链接

[API整理-知乎文章](https://zhuanlan.zhihu.com/p/587004798)

- markdown文件下载：[alist.md](https://github.com/alist-org/alist/files/10096470/alist.md)
- pdf文件下载：[alist.pdf](https://github.com/alist-org/alist/files/10096471/alist.pdf)

## 整理的API列表

- POST token获取 /api/auth/login
- POST 获取文件列表 /api/fs/list
- POST 新建文件夹  /api/fs/mkdir
- POST 重命名文件/目录  /api/fs/rename
- POST 删除文件/目录 /api/fs/remove
- PUT 表单上传文件  /api/fs/form
- POST 获取下载信息  /api/fs/get
- GET 系统设置  /api/admin/setting/list
- GET 账户列表  /api/admin/user/list
- GET 存储列表  /api/admin/storage/list
- POST 启用存储 /api/admin/storage/enable
- POST 禁用存储  /api/admin/storage/disable
- POST 新建存储  /api/admin/storage/create
- GET 查询指定存储信息 /api/admin/storage/get
- GET 查询所有驱动配置模板列表 /api/admin/driver/list
- POST 删除指定存储 /api/admin/storage/delete

## POST token获取

POST /api/auth/login

### 请求参数

|名称|位置|类型|必选|说明|
|---|---|---|---|---|
|Password|query|string| 是 |密码|
|Username|query|string| 是 |用户名|

> 返回示例

> 成功

```json
{
  "code": 200,
  "message": "success",
  "data": {
    "token": "abcd"
  }
}
```

### 返回结果

|状态码|状态码含义|说明|数据模型|
|---|---|---|---|
|200|[OK](https://tools.ietf.org/html/rfc7231#section-6.3.1)|成功|Inline|

### 返回数据结构

状态码 **200**

|名称|类型|必选|约束|中文名|说明|
|---|---|---|---|---|---|
|» code|integer|true|none||状态码|
|» message|string|true|none||信息|
|» data|object|true|none||data|
|»» token|string|true|none||token|

**Comment by chung1912** [comment]
{
    "code": 402,
    "message": "Invalid 2FA code",
    "data": null
}

开了二级验证的，应该怎么获取？

**Reply by Kuingsmile** [reply]
> { "code": 402, "message": "Invalid 2FA code", "data": null }
> 
> 开了二级验证的，应该怎么获取？
![image](https://user-images.githubusercontent.com/96409857/223303359-b2562586-ed13-4327-9fb5-cc8fd310a2eb.png)

![image](https://user-images.githubusercontent.com/96409857/223303262-9a470da1-e122-4d9e-8687-d438e6b41164.png)
请增加OtpCode参数，我这边测试可用

**Reply by chung1912** [reply]
请问topt（6个数字）的参数呢？有这个么？

**Reply by Kuingsmile** [reply]
也是这个吧，源码里这个接口就只有username，password，otpcode这三个参数，我用的google身份验证器，也是6个数字的

**Comment by chung1912** [comment]
请问API能监控正在下载或上传的任务吗？

**Reply by chung1912** [reply]
请问API能监控正在下载或上传的任务吗？

**Reply by Toserk1024** [reply]
复制：/api/admin/task/copy/undone
上传：/api/admin/task/upload/undone

**Reply by Toserk1024** [reply]
其他的你自己抓包一下就可以看到了

**Reply by Kuingsmile** [reply]
可以的 参考下面最新整理的api文档的task部分

alist中的任务管理可以分为：

| 接口名 | 说明 |
| --- | --- |
| /upload | 上传任务 |
| /copy | 复制任务 |
| /aria2_down | aria2下载任务 |
| /aria2_transfer | aria2转存任务 |
| /qbit_down | qbit下载任务 |
| /qbit_transfer | qbit转存任务 |

每种任务都有以下接口：

- done
- undone
- delete
- cancel
- clear_done
- clear_succeeded
- retry

[alist v3 api.md](https://github.com/alist-org/alist/files/12354885/alist.v3.api.md)

**Reply by pongfcnkl** [reply]
为什么我获取到的是null
{'code': 200, 'message': 'success', 'data': null}
复制里明明有任务的

**Comment by youtehub** [comment]
哈哈， 在此支持一下

**Comment by jsdfhasuh** [comment]
put 文件请求到 f"{host}/api/fs/form"
headers:{'Authorization': 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6ImpzZGZoYXN1aCIsImV4cCI6MTY5MTg5NTkxNCwibmJmIjoxNjkxNzIzMTE0LCJpYXQiOjE2OTE3MjMxMTR9.kkNrzM05OFLNPH8Sa5nqLYzS4xkc13aQIjfti7WsYpU', 'Content-Type': 'multipart/form-data;', 'File-Path': '%2Fshare%2FSRT_Store%2F%E6%96%B0%E5%BB%BA%E6%96%87%E6%9C%AC%E6%96%87%E6%A1%A3.srt', 'Content-Length': '130'}
data:{{file': <_io.BufferedReader name='C:\\Users\\dd109\\SynologyDrive\\python代码\\远程剧集同步\\新建文本文档.srt'>}
response = requests.put(url, data=files, headers=headers)

回应：'{"code":500,"message":"no multipart boundary param in Content-Type","data":null}'
请问大佬们，这该如何解决呢

**Reply by Kuingsmile** [reply]
tg群里回复你了哈

**Comment by anwen-anyi** [comment]
菜鸡的我，最近尝试了一下API 为啥这个重命名文件夹的不对呢~
`name`：是新需要命名的应该没错吧
`path`：是旧文件夹路径，路径最后结尾的是旧文件件名字？也没错吧
- 然后我反过来填写一下也不行，不知道那里错了QAQ，是有挂载这个文件夹的图中红色圈出来了

{{103}}是我设置了通用域名没问题的
![image](https://github.com/alist-org/alist/assets/56105412/1383c6a3-f233-4b02-8fd0-e6b12b38cb9f)

**Reply by Kuingsmile** [reply]
![image](https://github.com/alist-org/alist/assets/96409857/6a4b84d8-bd12-4aab-8a41-43bc556df9f0)
@anwen-anyi body不要用x-www-form-urlencoded,后端接收的是json数据，选raw，然后右边选json

**Reply by Kuingsmile** [reply]
然后header里要添加content-type为application/json
![image](https://github.com/alist-org/alist/assets/96409857/80bcc576-ad6b-4d7e-a668-a6836b814faf)

**Reply by anwen-anyi** [reply]
> ![image](https://user-images.githubusercontent.com/96409857/261211271-6a4b84d8-bd12-4aab-8a41-43bc556df9f0.png) @anwen-anyi body不要用x-www-form-urlencoded,后端接收的是json数据，选raw，然后右边选json

都什么时候用在 body选择 `raw` ？
我目前就是试试看，然后都是用的`x-www-form-urlencoded`，难怪不行 哈哈  谢谢了

**Reply by Kuingsmile** [reply]
除了get请求的接口都是body发送json数据 😄

**Reply by anwen-anyi** [reply]
> 除了get请求的接口都是body发送json数据 😄

好嘛，这次明白哩，我还傻鸟的特意搜索了x-www-form-urlencoded和form-data有什么区别哈哈，忘记了raw这个玩意... 谢谢了

**Comment by youtehub** [comment]
支持一下，要是能整合进官方文档就好了  
[https://alist.nn.ci/zh/guide/advanced/sso.html](https://alist.nn.ci/zh/guide/advanced/sso.html)

**Reply by anwen-anyi** [reply]
https://alist.nn.ci/zh/guide/api  已经合并了

**Comment by lsp0701** [comment]


**Reply by Kuingsmile** [reply]
https://alist.nn.ci/zh/guide/api/admin/task.html  copy接口的使用方式和task下的其它接口一致，参考task/upload接口就行
![image](https://github.com/alist-org/alist/assets/96409857/04e60e18-f230-496b-9794-e00f65e1dd9f)

**Reply by pongfcnkl** [reply]
https://alist.nn.ci/zh/guide/api/fs.html#post-%E5%A4%8D%E5%88%B6%E6%96%87%E4%BB%B6

**Comment by hejiadong0608** [comment]
有获取下载直链的接口吗？自己拼接 "/d + 路径 + sign" 的方式不一定准确，例如目前登陆了一个指定了子目录的账号的情况。

**Reply by Kuingsmile** [reply]
/api/fs/get接口拿到的数据里有raw_url字段，把这个字段的/p改成/d就行，我试了下指定子目录的账号拿到的路径也是对的

**Reply by hejiadong0608** [reply]
那是因为你的地址是代理地址吧，302跳转的这个字段里就是官方的直链了。

**Comment by Shunamxiao** [comment]
大佬你是用哪个接口工具导出的啊

**Reply by Kuingsmile** [reply]
APIFox

**Reply by Shunamxiao** [reply]
有个问题，/api/fs/form表单上传接口中，哪个字段是表示上传到哪个目录的啊，我按照提示返回说没有当前存储

**Reply by Shunamxiao** [reply]
File-Path	header	string	是		经过URL编码的完整文件路径
这个是url编码的文件绝对路径？
file也是文件了

**Reply by Kuingsmile** [reply]
![image](https://github.com/alist-org/alist/assets/96409857/feaf6e28-970c-49c1-af85-a460e4d21809)
![image](https://github.com/alist-org/alist/assets/96409857/5128a527-ed6f-407f-b98a-b0b01b3dbe84)
类似这样

**Reply by Shunamxiao** [reply]
storage not found; please add a storage first

**Reply by Shunamxiao** [reply]
已解决，文件大小要注意

**Comment by hankqqu** [comment]
> put 文件请求到 f"{host}/api/fs/form"
> headers:{'Authorization': 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6ImpzZGZoYXN1aCIsImV4cCI6MTY5MTg5NTkxNCwibmJmIjoxNjkxNzIzMTE0LCJpYXQiOjE2OTE3MjMxMTR9.kkNrzM05OFLNPH8Sa5nqLYzS4xkc13aQIjfti7WsYpU', 'Content-Type': 'multipart/form-data;', 'File-Path': '%2Fshare%2FSRT_Store%2F%E6%96%B0%E5%BB%BA%E6%96%87%E6%9C%AC%E6%96%87%E6%A1%A3.srt', 'Content-Length': '130'}
> data:{{file': <_io.BufferedReader name='C:\\Users\\dd109\\SynologyDrive\\python代码\\远程剧集同步\\新建文本文档.srt'>}
> response = requests.put(url, data=files, headers=headers)
> 
> 回应：'{"code":500,"message":"no multipart boundary param in Content-Type","data":null}'
> 请问大佬们，这该如何解决呢

**Comment by xlingsy** [comment]
POST复制文件
> /api/fs/copy

和POST移动文件
> /api/fs/move

今天试这两个api，发现返回的结果都是
> {'code': 400, 'message': 'Empty file names', 'data': None}

但是请求data里面明明是有names参数的
>  {'src_dir': '/temp/backup', 'dst_dir': '/temp', 'names': ['alist', 'aria2', 'ntfy']}

**Reply by dream10201** [reply]
看源码，是因为需要一个json参数
```python
requests.post(f"{URL}/api/fs/copy",headers=self.headers,json=data) => ok
requests.post(f"{URL}/api/fs/copy",headers=self.headers,data=data) => ng
```

**Comment by aiaidu** [comment]
请问我是用fs/search搜索文件，返回结果中total是1但是contents中为空是什么原因
